src/gluonnlp/data/datasetloader.py

```diff
@@ -0,0 +1,424 @@
+    dataset_params : dict, default is None
+        Dictionary of parameters passed to dataset_fn.
+    batch_sampler_params : dict, default is None
+        Dictionary of parameters passed to batch_sampler_fn.
+    batchify_fn : callable
+        Callback function to allow users to specify how to merge samples
+        into a batch. Defaults to `default_batchify_fn`::
+
+            def default_batchify_fn(data):
+                if isinstance(data[0], nd.NDArray):
+                    return nd.stack(*data)
+                elif isinstance(data[0], tuple):
+                    data = zip(*data)
+                    return [default_batchify_fn(i) for i in data]
+                else:
+                    data = np.asarray(data)
+                    return nd.array(data, dtype=data.dtype)
+    num_dataset_workers : int
+        Number of worker process for dataset creation.
+    num_batch_workers : int
+        Number of worker process for batch creation.
+    pin_memory : boolean, default False
+        If ``True``, the dataloader will copy NDArrays into pinned memory
+        before returning them. Copying from CPU pinned memory to GPU is faster
+        than from normal CPU memory.
+    circle_length : int, default is 1
+        The number of files to be read at the same time. When circle_length is larger than 1,
+        we merge circle_length files.
+    dataset_prefetch : int, default is `num_dataset_workers`
+        The number of prefetching datasets only works if `num_workers` > 0.
+        If `prefetch` > 0, it allow worker process to prefetch certain datasets before
+        acquiring data from iterators.
+        Note that using large prefetching batch will provide smoother bootstrapping performance,
+        but will consume more memory. Using smaller number may forfeit the purpose of using
+        multiple worker processes, try reduce `num_dataset_workers` in this case.
+        By default it defaults to `num_dataset_workers`.
+    batch_prefetch : int, default is `num_batch_workers * 2`
+        The number of prefetching batches only works if `num_workers` > 0.
+        If `prefetch` > 0, it allow worker process to prefetch certain batches before
+        acquiring data from iterators.
+        Note that using large prefetching batch will provide smoother bootstrapping performance,
+        but will consume more shared_memory. Using smaller number may forfeit the purpose of using
+        multiple worker processes, try reduce `num_batch_workers` in this case.
+        By default it defaults to `num_batch_workers * 2`.
+    dataset_cached : bool, default is False
+        Whether or not to cache last processed dataset. Each processed dataset can
+        only be cached for once. When there is no new available processed dataset to be fetched,
+        we pop a cached processed dataset.
+    num_max_dataset_cached : int, default is 0
+        Maximum number of cached datasets. It is valid only if dataset_cached is True
```

Comment:
This class is fairly complicated for a new user to understand. Do we plan to have example in this doc, or adding it to the `api_notes/data.rst` to help users understand how to use this for large scale dataset?

Reply:
Of course, I think we can use bert training as an example.